### ISSUE TITLE
False positive in unnecessary_cast with hex literals

### ISSUE THREAD

**Issue by glfmn** [rust-lang/rust-clippy#5220]
clippy 0.0.212 (69f99e7 2019-12-14)

The lint for `#[warn(unnecessary_cast)]` acts strangely when integers are written in hex.  Example:

```
let r = 0x1b as f32 / 255.0;
//      ^^^^^^^^^^^ - unnecessary cast write integer literal as 29_f32
```

However, `f32` is valid hex, so if I write `0x1b_f32` rust interprets the value as an integer.  It _does_ helpfully output the integer literal as a decimal number in the suggestion, but it isn't hex anymore which defeats the purpose.

**Comment by mlegner** [comment]
I would like to work on this issue and investigated a bit. Unfortunately, the AST doesn't seem to record whether the integer literal was written in hex or decimal notation. The enum [`LitIntType`](https://github.com/rust-lang/rust/blob/master/src/librustc_ast/ast.rs#L1506) only records the suffix of the literal.

It seems to me that properly resolving these false positives would require extending the [`LitKind::Int`](https://github.com/rust-lang/rust/blob/master/src/librustc_ast/ast.rs#L1539) node with additional information in the main [rust](https://github.com/rust-lang/rust) repository. Am I missing something?

**Comment by glfmn** [comment]
@mlegner that's unfortunate, I'm not really aware of those implementation details but from a glance it seems your summation of the problem is accurate.

**Comment by flip1995** [comment]
Clippy has it's own struct, that can be build from literals and from where you can detect if it is a hex, bin, or oct literal:

https://github.com/rust-lang/rust-clippy/blob/0f4a3fecccd55c0e5b4d50ce59ed660ac8e67dcc/clippy_lints/src/literal_representation.rs#L132-L150

So no need to modify the rustc AST.

**Comment by mlegner** [comment]
Thanks a lot for the pointer, @flip1995. I was able to resolve this issue but had to modify some of the code you referenced. The changes are in the linked PR  #5257.

**Comment by glfmn** [comment]
Thanks for addressing this so quickly!
